regex-parsing/src/main/java/org/sonarsource/analyzer/commons/regex/finders/VerboseRegexFinder.java

```diff
@@ -0,0 +1,167 @@
+import static org.sonarsource.analyzer.commons.regex.ast.CharacterClassElementTree.Kind.ESCAPED_CHARACTER_CLASS;
+import static org.sonarsource.analyzer.commons.regex.ast.CharacterClassElementTree.Kind.PLAIN_CHARACTER;
+import static org.sonarsource.analyzer.commons.regex.ast.CharacterClassElementTree.Kind.UNION;
+
+public class VerboseRegexFinder extends RegexBaseVisitor {
+
+  private static final String VERBOSE_QUANTIFIER_MESSAGE = "Use concise quantifier syntax '%s' instead of '%s'.";
+  private static final String VERBOSE_CHARACTER_CLASS_MESSAGE = "Use concise character class syntax '%s' instead of '%s'.";
+
+  private final RegexIssueReporter.ElementIssue regexElementIssueReporter;
+
+  public VerboseRegexFinder(RegexIssueReporter.ElementIssue regexElementIssueReporter) {
+    this.regexElementIssueReporter = regexElementIssueReporter;
+  }
+
+
+  @Override
+  public void visitCharacterClass(CharacterClassTree tree) {
+    checkBulkyAlphaNumericCharacterClass(tree);
+    checkBulkyNumericCharacterClass(tree);
+    checkBulkyAnyCharacterClass(tree);
+    super.visitCharacterClass(tree);
+  }
+
+  @Override
+  public void visitRepetition(RepetitionTree tree) {
+    checkBulkyQuantifier(tree.getQuantifier());
+
+    super.visitRepetition(tree);
+  }
+
+  private void checkBulkyAlphaNumericCharacterClass(CharacterClassTree tree) {
+    CharacterClassElementTree element = tree.getContents();
+    if (element.is(UNION) && ((CharacterClassUnionTree) element).getCharacterClasses().size() == 4) {
+      boolean hasDigit = false;
+      boolean hasLowerCase = false;
+      boolean hasUpperCase = false;
+      boolean hasUnderscore = false;
+
+      for (CharacterClassElementTree subElement : ((CharacterClassUnionTree) element).getCharacterClasses()) {
+        String raw = subElement.getText();
+        if (subElement.is(CHARACTER_RANGE)) {
+          hasDigit |= "0-9".equals(raw);
+          hasLowerCase |= "a-z".equals(raw);
+          hasUpperCase |= "A-Z".equals(raw);
+        } else if (subElement.is(PLAIN_CHARACTER)) {
+          hasUnderscore |= "_".equals(raw);
+        }
+      }
+
+      if (hasDigit && hasLowerCase && hasUpperCase && hasUnderscore) {
+        String expected = backslash(tree) + (tree.isNegated() ? "W" : "w");
+        reportVerboseCharacterClass(expected, tree);
+      }
+    }
+  }
+
+  private void checkBulkyNumericCharacterClass(CharacterClassTree tree) {
+    CharacterClassElementTree element = tree.getContents();
+    if (element.is(CHARACTER_RANGE) && "0-9".equals(element.getText())) {
+      String expected = backslash(tree) + (tree.isNegated() ? "D" : "d");
+      reportVerboseCharacterClass(expected, tree);
+    }
+  }
+
+  private void checkBulkyAnyCharacterClass(CharacterClassTree tree) {
+    CharacterClassElementTree element = tree.getContents();
+    if (tree.isNegated() || !element.is(UNION) || ((CharacterClassUnionTree) element).getCharacterClasses().size() != 2) {
```

Comment:
I was gonna ask "What about `/[\w\Wx]/` ?"
But then this is probably going to fire another rule, stating that the 'x' is redundant, and when 'x' is removed the rest will lend in this case?

Reply:
It will be covered by the rule which detects duplications in character classes